File: src/investments/registered-education-savings-plan/tuition-fees.ts
/*
 Sources
 https://www150.statcan.gc.ca/t1/tbl1/en/tv.action?pid=3710004501&cubeTimeFrame.startYear=2024+%2F+2025&cubeTimeFrame.endYear=2024+%2F+2025&referencePeriods=20240101%2C20240101

 Last updated
 2025-12-21
 */

import { ByProvince, ProvinceCode } from '../../misc';

export interface TuitionFees {
    TuitionFeesData: ByProvince<number>;
    getTuitionFeesByProvinceCode(provinceCode: ProvinceCode): number;
}

const TuitionFeesData: ByProvince<number> = {
    AB: 8067,
    BC: 6862,
    MB: 5993,
    NB: 9938,
    NL: 3746,
    NS: 9988,
    PE: 8191,
    ON: 8958,
    QC: 3963,
    SK: 9863,
    NT: 7734, // using geography = Canada
    NU: 7734, // using geography = Canada
    YT: 4470,
};

const getTuitionFeesByProvinceCode = (provinceCode: ProvinceCode): number => TuitionFeesData[provinceCode];

export const TuitionFees: TuitionFees = {
    getTuitionFeesByProvinceCode,
    TuitionFeesData,
};

File: src/commonMain/kotlin/com/equisoft/taxca/investments/resp/TuitionFees.kt
/*
 Sources
 https://www150.statcan.gc.ca/t1/tbl1/en/tv.action?pid=3710004501&cubeTimeFrame.startYear=2024+%2F+2025&cubeTimeFrame.endYear=2024+%2F+2025&referencePeriods=20240101%2C20240101

 Last updated
 2025-12-21
 */

package com.equisoft.taxca.investments.resp

import com.equisoft.taxca.misc.Jurisdiction

// The TS source merges an interface and a const named `TuitionFees`; Kotlin forbids a
// top-level class and val sharing a name, so the class carries a `Resp` prefix.
data class RespTuitionFees(
    val tuitionFeesData: Map<Jurisdiction, Double>,
) {
    fun getTuitionFeesByProvinceCode(provinceCode: Jurisdiction): Double =
        tuitionFeesData.getValue(provinceCode)
}

val TuitionFees: RespTuitionFees = RespTuitionFees(
    tuitionFeesData = mapOf(
        Jurisdiction.AB to 8067.0,
        Jurisdiction.BC to 6862.0,
        Jurisdiction.MB to 5993.0,
        Jurisdiction.NB to 9938.0,
        Jurisdiction.NL to 3746.0,
        Jurisdiction.NS to 9988.0,
        Jurisdiction.PE to 8191.0,
        Jurisdiction.ON to 8958.0,
        Jurisdiction.QC to 3963.0,
        Jurisdiction.SK to 9863.0,
        // using geography = Canada
        Jurisdiction.NT to 7734.0,
        // using geography = Canada
        Jurisdiction.NU to 7734.0,
        Jurisdiction.YT to 4470.0,
    ),
)
